File: javascript/ChineseCheckers.js
class Tile {
    constructor(x, y, z) {
        //hexagonal coordinate system
        //lattice points of plane x + y + z = 0
        //project z to plane if not on plane
        if (x + y + z != 0) {
            z = - x - y;
        }
        this.x = x;
        this.y = y;
        this.z = z;
    }
    toString() {
        return `${this.x}, ${this.y}, ${this.z}`;
    }
    //vector operations
    equals(that) {
        return (this.x === that.x) && (this.y === that.y);
    }
    add(that) {
        return new Tile(this.x + that.x, this.y + that.y, this.z + that.z);
    }
    sub(that) {
        return new Tile(this.x - that.x, this.y - that.y, this.z - that.z);
    }
    neg() {
        return new Tile(-this.x, -this.y, -this.z);
    }
    //scalar multiplication
    mult(that) {
        return new Tile(this.x * that, this.y * that, -this.z * that);
    }
    //rectilinear distance (adjacent tiles have distance 1)
    dist(that) {
        return (Math.abs(this.x - that.x) + Math.abs(this.y - that.y) + Math.abs(this.z - that.z))/2;
    }
    //hexagonal coordinates to 2D rectangular for board display
    //side (0-5) for all 6 rotation viewpoints
    //yay for linear algebra!
    to_rect(side = 0) {
        const c = [2, 1, -1, -2, -1, 1][side];
        const s = [0, 1, 1, 0, -1, -1][side];
        const x = (c + 3 * s) * this.x + (-c + 3 * s) * this.y;
        const y = (s - c) * this.x + (-s - c) * this.y;
        return [x/2, y/2] //normalized to integers for easy server-client transmission, multiply y coordinate by sqrt(3) = ~1.73 before display.
    }
}

class TileSet {
    constructor(tiles = null) {
        if (tiles === null) {
            tiles = []
        }
        this.map = new Map()
        tiles.forEach(element => {
            this.map.set(element.toString(),element)
        });
    }
    add(tile) {
        this.map.set(tile.toString(), tile);
    }
    values() {
        return this.map.values();
    }
    delete(tile) {
        return this.map.delete(tile.toString());
    }
    update(other) {
        [...other.values()].forEach(element => {
            this.add(element)
        });
    }
    clear() {
        this.map.clear()
    }
    difference(other) {
        return new TileSet([...(this.map.values())].filter(tile => !other.map.has(tile.toString())))
    }
    copy() {
        return new TileSet([...this.map.values()])
    }
    includes(tile) {
        return this.map.has(tile.toString)
    }
}
//for (let index = 0; index < 6; index++) {
//    console.log(tile.toRect(index));
//}

class ChineseCheckers {
    constructor() {
        //Map of all pieces and tile locations, color as key
        this.pieces = new Map()
        //Map of home side index 0-5
        //used to check win condition
        this.home = new Map()

        //six cardinal directions
        //for getting adjacent tiles
        this.directions = [new Tile(1,0,-1), new Tile(0,1,-1), new Tile(-1,0,1), new Tile(0,-1,1), new Tile(1,-1,0), new Tile(-1,1,0)]
        this.isPlaying = false

    }
    set_up(gamemode = 2, colors = ['red','yellow','green','cyan','blue','magenta']) {
        //gamemode number of players; 2, 3, 4 or 6 for symmetry
        //default color wheel colors
        this.reset()
        for (let index = 0; index < gamemode; index++) {
            const i = Math.floor(index*6/gamemode);
            const color = colors[i];
            let piecearray = [];
            const c = [1,0,0,-1,0,0][i % 6];
            const d = [1,0,0,-1,0,0][(i + 2) % 6];
            const e = [1,0,0,-1,0,0][(i + 4) % 6];
            for (let a = 1; a < 5; a++) {
                for (let b = 5 - a; b < 5; b++) {
                    const x = c * a + d * b + e * (-a - b);
                    const y = e * a + c * b + d * (-a - b);
                    piecearray.push(new Tile(x, y, -x-y));
                }
            }
            this.pieces.set(color, piecearray);
            this.home.set(color, index);
        this.isPlaying = true;
        }
    }
    //calculates home tile from index 0-5.
    home_tile(i) {
        const c = [1,0,0,-1,0,0][i % 6];
        const d = [1,0,0,-1,0,0][(i + 2) % 6];
        const e = [1,0,0,-1,0,0][(i + 4) % 6];
        const x = c * 4 + d * 4 + e * (-8);
        const y = e * 4 + c * 4 + d * (-8);
        return new Tile(x, y, -x-y);
    }
    reset() {
        this.pieces.clear;
        this.home.clear;
        this.isPlaying = false;
    }
    //boundary is intersection of two equilateral triangles
    in_bounds(tile) {
        return (tile.x <= 4 && tile.y <= 4 && tile.z <= 4) || (tile.x >= -4 && tile.y >= -4 && tile.z >= -4);
    }
    occupied_tiles() {
        let occupied = [];
        for (const colorpieces of [...this.pieces.values()]) {
            occupied.push(...colorpieces)
        }
        return occupied;
    }
    is_occupied(tile) {
        return !!this.occupied_tiles().find(element => element.equals(tile));
    }
    is_empty(tile) {
        return !(this.is_occupied(tile));
    }
    move(loc, dest) {
        for (const color in this.pieces) {
            for (let i = 0; i < this.pieces[color].length; i++) {
                if (loc.equals(this.pieces[color][i])) {
                    this.pieces[color][i] = dest
                }
            }
        }
    }
    //find all tile connected by jump move
    connection(tile) {
        let aggr = new TileSet();
        let newtiles = new TileSet([tile]);
        while (newtiles.map.size > 0) {
            aggr.update(newtiles);
            for (const x of newtiles.copy().values()) {
                for (const direction of this.directions) {
                    let jump = x.add(direction.mult(2));
                    console.log(jump.toString())
                    if (this.is_occupied(x.add(direction)) && this.in_bounds(jump) && this.is_empty(jump)) {
                        newtiles.add(jump);
                        console.log('added')                      
                    }
                }
            }
            newtiles = newtiles.difference(aggr);
        }
        return aggr;
    }
    //determines if move is legal
    //move must be to adjacent tile, or connected by jump
    is_legal(loc, dest) {
        if (this.is_occupied(loc)) {
            if (this.is_empty(dest) && this.in_bounds(dest)) {
                if (loc.dist(dest) === 1) {
                    return [true, 'step']
                }
                else{
                    if (this.connection(loc).includes(dest) && !loc.equals(dest)) {
                        return [true, 'jump']
                    } else {
                        return [false, 'no path found']
                    }
                }
            } else {
                return [false, 'destination occupied or out of bound']
            }
        } else {
            return [false, 'no piece to move']
        }
    }
    //win condition
    has_won(color) {
        home = this.home_tile(this.home.get(color))
        return this.pieces.get(color).every(piece => piece.distance(home) > 12)
    }
    //all pieces mapped to rectangular coordinates
    to_rect(side = 0) {
        let pieces_rect = new Map()
        for (const color of this.pieces.keys()) {
            pieces_rect.set(color, this.pieces.get(color).map(tile => tile.to_rect(side)));
        }
        return pieces_rect
    }
}

File: python/ChineseCheckers.py
#!/usr/bin/env python
# coding: utf-8
import turtle
class Tile:
    def __init__(self, x, y, z):
        #Hexagonal Coordinate System. Uses lattice points (x,y,z) confined to the plane x+y+z = 0
        #
        #e.g.:
        #    (0,-1,1) (-1,0,1)
        #
        #(1,-1,0) (0,0,0) (-1,1,0)
        #
        #    (1,0,-1) (0,1,-1)
        if x + y + z != 0: #projects z to the plane if not on plane.
            z = -x - y
        
        self.x = x
        self.y = y
        self.z = z

    def __str__(self):
        return '({}, {}, {})'.format(self.x, self.y, self.z)
    
    def __repr__(self):
        return str(self)

    def triple(self):
        return self.x, self.y, self.z
    
    #Hash:
    def __hash__(self):
        return hash((self.x,self.y))
        #Every Tile is uniquely determined by its x and y coordinate, so that tuple is used for the hash.
        #This is necessary for creating sets of Tiles.
    
    #Possibly Useful Operations (essentially vector operations):
    def __eq__(self, other):
        return (self.x == other.x) and (self.y == other.y)
    
    def __sub__(self, other):
        return Tile(self.x - other.x, self.y - other.y, self.z-other.z)
    
    def __neg__(self):
        return Tile(-self.x, -self.y, -self.z)
    
    def __add__(self, other):
        return Tile(self.x - other.x, self.y - other.y, self.z-other.z)
    
    #Scalar Multiplication:
    def __mul__(self, other):
        return Tile(self.x * other, self.y * other, self.z * other)
    
    #Rectilinear or 'Taxicab' Distance. Adjacent tiles have distance 1:
    def distance(self, other): 
        return (abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z-other.z))//2

    #Returns the coordinates of the hexagonal coordinate system mapped to a 2D plane.
    def rect(self, index = 2, sgn = -1):
         
        x = sgn*(self.x, self.y, self.z)[(index+1)%3]
        y = sgn*(self.y, self.z, self.x)[(index+1)%3]
        
        return (x-y, x+y) #second coordinate should be multiplied by sqrt 3 when drawn to preserve lengths.

    
class ChineseCheckers:
    def __init__(self):
        #Pieces and location stored in dict in the form {color: [tile1,tile2,tile3,...,tile10]}
        self.pieces = {}
        
        #Home tile is the vertex in the starting triangle farthest from center. In the form {color: tile}
        #used to check for the win condition
        self.home = {}
        
    
        #Six cardinal directions:
        #useful for adding to tile to get adjacent tiles.
        self.directions = [Tile(1,0,-1),Tile(0,1,-1),Tile(-1,0,1),Tile(0,-1,1),Tile(1,-1,0),Tile(-1,1,0)]
        self.playing = False
        self.drawn = False
        
        
    def set_up(self, gamemode = 2):
        self.reset()
        #Different setups (gamemode is number of players/colors, default 2):
        if gamemode == 2:
            self.pieces['red'] = []
            self.pieces['blue'] = []
            for x in range(1,5):
                for y in range(5-x,5):
                    self.pieces['red'].append(Tile(x,y,-x-y))
                    self.pieces['blue'].append(Tile(-x,-y,x+y))
            self.home['red'] = Tile(4,4,-8)
            self.home['blue'] = Tile(-4,-4,8)
            
        if gamemode == 3:
            self.pieces['red'] = []
            self.pieces['blue'] = []
            self.pieces['green'] = []
            for x in range(1,5):
                for y in range(5-x,5):
                    self.pieces['red'].append(Tile(x,y,-x-y))
                    self.pieces['blue'].append(Tile(-x-y,x,y))
                    self.pieces['green'].append(Tile(y,-x-y,x))
            self.home['red'] = Tile(4,4,-8)
            self.home['blue'] = Tile(-8,4,4)
            self.home['green'] = Tile(4,-8,4)
            
        if gamemode == 4:
            self.pieces['red'] = []
            self.pieces['blue'] = []
            self.pieces['green'] = []
            self.pieces['yellow'] = []
            for x in range(1,5):
                for y in range(5-x,5):
                    self.pieces['red'].append(Tile(x,y,-x-y))
                    self.pieces['blue'].append(Tile(-x,-y,x+y))
                    self.pieces['green'].append(Tile(y,-x-y,x))
                    self.pieces['yellow'].append(Tile(-y,x+y,-x))
            self.home['red'] = Tile(4,4,-8)
            self.home['blue'] = Tile(-4,-4,8)
            self.home['green'] = Tile(4,-8,4)
            self.home['yellow'] = Tile(-4,8,-4)
        
        if gamemode == 6:
            self.pieces['red'] = []
            self.pieces['blue'] = []
            self.pieces['green'] = []
            self.pieces['yellow'] = []
            self.pieces['cyan'] = []
            self.pieces['magenta'] = []
            for x in range(1,5):
                for y in range(5-x,5):
                    self.pieces['red'].append(Tile(x,y,-x-y))
                    self.pieces['blue'].append(Tile(-x,-y,x+y))
                    self.pieces['green'].append(Tile(y,-x-y,x))
                    self.pieces['yellow'].append(Tile(-y,x+y,-x))
                    self.pieces['cyan'].append(Tile(-x-y,x,y))
                    self.pieces['magenta'].append(Tile(x+y,-x,-y))
            self.home['red'] = Tile(4,4,-8)
            self.home['blue'] = Tile(-4,-4,8)
            self.home['green'] = Tile(4,-8,4)
            self.home['yellow'] = Tile(-4,8,-4)
            self.home['cyan'] = Tile(-8,4,4)
            self.home['magenta'] = Tile(8,-4,-4)
        self.playing = True
                    
    def reset(self):
        self.playing = False
        self.pieces = {}
        self.home = {}
     
    def __repr__(self):
        return str(self.pieces)
    
    #Mathematically defined boundaries, the union of two large equilateral triangles to get hexagram:
    def in_bounds(self, tile):
        return (tile.x <= 4 and tile.y <= 4 and tile.z <= 4) or (tile.x >= -4 and tile.y >= -4 and tile.z >= -4)
    
    #Useful test functions:
    def occupied_tiles(self):
        occupied = []
        for x in self.pieces.values():
            occupied.extend(x)
        return occupied
    
    def is_occupied(self, tile):
        return tile in self.occupied_tiles()
    
    def is_empty(self, tile): 
        return tile not in self.occupied_tiles()
        
    #Moving pieces:
    #first tests if move is legal (see below)
    def move(self, loc, dest):
        if self.in_bounds(dest):
            for color in self.pieces:
                for i in range(len(self.pieces[color])):
                    if self.pieces[color][i] == loc:
                        self.pieces[color][i] = dest
    
    #Finding all possible connected 'jumps' from a given tile:
    #builds the set of connected tiles (ensures no duplication)
    #keeps checking all 6 jump directions
    #only checks 'new' adjacent tiles for efficiency
    #terminates when no new tiles are found
    def connection(self, tile):
        connected = set()
        new = set([tile])
        
        while len(new) > 0:
            connected.update(new)
            for x in new.copy():
                for direction in self.directions:
                    if self.is_occupied(x + direction) and self.in_bounds(x+direction*2) and self.is_empty(x+direction*2):
                        new.add(x+direction*2)
            new = new - connected
        
        return connected
    
    #Test for legal moves:
    #move must be to adjacent tile or tile connected by jumps
    def is_legal(self, loc, dest):
        if self.is_occupied(loc):
            if self.is_empty(dest) and self.in_bounds(dest):
                if loc.distance(dest)==1:
                    return True, 'step'
                elif dest in self.connection(loc)-set([loc]):
                    return True, 'jump'
                
                else:
                    return False, 'no path found'
            else:
                return False, 'destination occupied or out of bound'
                
        else:
            return False, 'no piece to move'
    
    #Win condition:
    #if all pieces are more than distance 12 from start, they must be in the opposite triangle, assuming they are in bounds.
    def has_won(self, color):
        return all([piece.distance(self.home[color]) > 12 for piece in self.pieces[color]])

    #Returns a dictionary of pieces with in rectangular coordinates remember multiply second coordinate by sqrt(3) to preserve length.
    #Send this to client after a move. Client draw on Tkinter canvas.
    def to_plane(self, home):
        for i in range(3):
            if abs(home.triple()[i]) == 8:
                index = i
                sgn = home.triple()[i]//8

        self.plane = {}
        for color in self.pieces:
            self.plane[color] = [tile.rect(index, sgn) for tile in self.pieces[color]]
        return self.plane

    #Test drawing with turtle graphics.
    def draw(self, home):
        if not self.drawn:
            self.turtle = turtle.Turtle()
            self.turtle.speed(10)
            self.turtle.hideturtle()
            self.drawn = True
        pieces = self.to_plane(home)
        self.turtle.clear()
        for color in pieces:
            for x in pieces[color]:
                self.turtle.penup()
                self.turtle.goto(20*x[0], 20*1.73*x[1])
                self.turtle.pendown()
                self.turtle.dot(25,color)
                
            
            
        
        








#Example
'''
C = ChineseCheckers()
print(C.pieces['red'])
print(C.connection(C.pieces['red'][4]))
C.move(Tile(3, 3, -6),Tile(3, 1, -4))


print(C.pieces['red'][4])
print(C.has_won('red'))
print(C.pieces['red'][4].distance(C.home['red']))
'''




